### DISCUSSION TITLE
Validation fails

### DISCUSSION THREAD

**Discussion by s-ohnishi** [phalcon/cphalcon#16057] [General]
I'm trying v5.0.0RC4 with PHP8.1.
In the class that extends \Phalcon\Forms\Form, we specified validation when defining the element.
```
            $elm->addValidators([
                new Validator\Digit([
                    'message' => 'Invalid number of reservations',
                    'allowEmpty' => false,
                ]),
                new Validator\Between([
                    'minimum' => 0,
                    'maximum' => 15
                ])
            ]);
```
I expect it to be a number from 0 to 15, but I want to make an error for other numbers, strings, and unset (null or '').
But with the above configuration, I get this error.
`Deprecated: ctype_digit(): Argument of type null will be interpreted as string in the future`

This is probably because the specification of ctype_*() has changed since PHP8.1, but Phalcon currently does not support it.

It would allow an unexpected value, but at the moment it's unavoidable, so even if I change it to Validator\Numericality, it won't recognize the posted value.
After trying various things, it seems that there is a problem with the name of the element.

I receive the form data in an array because it takes less time to process it after it is received.
For example:
```
<input type="number" name="foo[1]" value="" />
<input type="number" name="foo[2]" value="" />
<input type="number" name="foo[3]" value="" />
```
It is often used in select/option instead of input.
(There are cases where multiple checkboxes are collectively acquired as `name="cb[]"`)
So I write the following in form::initialize().
```
        $optionAry = array(0,1,2,3);
        $elm = new Element\Select("order[1]", $optionAry, [
            'id' => 'order11',
            'useEmpty' => false,
        ]);
```
With this, you can get the HTML element with the name with [ ] in the view, but it doesn't seem to understand it during validation.
If you change "order[1]" to "order1", it will be processed without problems, but considering the processing after receiving it, I would like to receive it as an array.
Is it a problem that you can set it and output HTML tags, but you can't validate it?

**Comment by niden** [comment]
I am pretty sure that arrays are a problem with form data - I have seen in the past people reporting it and have not seen an implementation for this. It is of course in the todo list because many people work with that.

As for the `ctype_` yes that is a bug we need to address.

**Reply by s-ohnishi** [reply]
It seems that `ctype_` has been resolved.
Thank you for your hard work.
I think it would be nice if another problem, the inability to verify elements with array type names, could be resolved.
I know you're busy, so I don't want you to make it your top priority.

**Reply by niden** [reply]
It is definitely on the todo list.

**Reply by s-ohnishi** [reply]
got it.
Looking forward to seeing it fixed.
